lec07-pop-models.Rmd

```diff
@@ -1,4 +1,617 @@
+  # K: carrying capacity
+  # N: population size
+  return(r * N * (1 - N / K))
+  }
+
+# parameters
+K <- 150; r <- 1  # you can put things on the same line, separated by a semicolon
+
+dt <- 0.05 #timestep - the smaller, the better
+tmax <- 8 # the total time we want to numerically solve for
+points <- tmax/dt # the number of data points in the simulation - add 1 so that we can start at t=0
+
+# vectors to store values of N and t at each timestep:
+N_vector <- numeric(points) # population size
+t_vector <- seq(0, tmax - dt, by = dt) # time vector
+
+# initial condition
+N0 <- 10
+N_vector[1] <- N0 
+
+N <- N0 # initialize variable N
+
+for (i in 2:points){ # start at 2 because the initial state is at position 1
+  dN <- logistic_eqn(N = N, r = r, K = K)*dt
+  N <- N + dN # the variable N is changing at each step of the loop
+  N_vector[i] <- N
+}
+
+qplot(t_vector, N_vector,
+      ylab = "Population size N",
+      xlab = "Time t") 
+```
+
+This numerical solution shows us that the population size grows until it reaches the carrying capacity $K$, the stable fixed point. 
+
+### A note on why $dt$ has to be small
+  
+If you're numerically solving a differential equation with Euler's method, $dt$ must be 'small'. 
+How small is small, and why does it have to be small? The answers to these are related, and to see why,
+let's look at what can happen if $dt$ is too big with an example from the logistic equation. 
+
+```{r}
+# what happens if we make dt too large when simulating? 
+# in this example, the starting population N0 is much larger than the carrying capacity.
+# This means dN/dt will be a large negative number: N will decrease quickly from such a large size.
+
+r <- 0.5
+K <- 50
+dt <- 1
+N0 <- 400
+
+dN <- logistic_eqn(N0, r, K)*dt # this is the change in N that we wil add to N0 to calculate N1
+
+N1 <- N0 + dN
+
+print(N1) # this is no good: N has shot from +400 to -1000 in a single step! 
+
+dN <- logistic_eqn(N1, r, K)*dt # calculate what the next change in N would be
+
+N2 <- N1 + dN
+
+print(N2) # N is getting more and more negative!
+ 
+```
+
+### Challenge - in groups
+
+Use Euler's method to calculate a numerical solution for the following differential equation:
+
+$$\frac{dx}{dt} = -2.3 x$$
+
+  - Try a few different step sizes ($\Delta t$) such as $0.7$ and $1$. What happens if $\Delta t$ is too large?
+
+This is what's called **numerical instability** --- the numerical solution grows very large 
+(either positive or negative) but the true solution doesn't. In the logistic equation example above,
+this numerical instability can happen if $N$ falls below $0$: the population will keep getting
+more and more negative after that. 
+
+## Using `deSolve` in R to solve differential equations numerically
+
+R has a package called `deSolve` which contains several functions to numerically solve differential equations.
+Let's look at the documentation for `deSolve`:
+
+```{r}
+library(deSolve)
+help(deSolve)
```

Comment:
This command shouldn't be used in a rmarkdown document. You can put the code chunk so it doesn't evaluate:

    ```{r}
    library(deSolve)
    ```
    ```{r, eval=FALSE}
    help(deSolve)
    ```